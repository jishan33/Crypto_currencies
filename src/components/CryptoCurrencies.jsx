import React, { useState, useContext } from "react";
import { CryptoCurrenciesContext } from ".././context/CryptoCurrenciesContext";
import DatePicker from "react-datepicker";
import moment from "moment";

import "react-datepicker/dist/react-datepicker.css";

const CryptoCurrencies = (props) => {
  const [context, setContext] = useContext(CryptoCurrenciesContext);
  const [cryptoCurrencies, setCryptoCurrencies] = useState([]);
  const [selectedDate, setSelectedDate] = useState(
    new Date("2019-12-04T00:00:00")
  );

  const formattedDate = (date, addDays) => {
    let response = moment(date, "YYYY-MM-DD").format("YYYY-MM-DD");
    if (addDays) {
      const afterDate = moment(date, "DD-MM-YYY").add(addDays, "day")._d;
      response = moment(afterDate).format("YYYY-MM-DD");
    }
    return response;
  };

  const getCryptoCurrencies = async () => {
    try {
      const date = formattedDate(selectedDate);
      const response = await fetch(
        `http://localhost:5000/crypto_currencies?date=${date}`
      );
      const data = await response.json();

      await setCryptoCurrencies(data);
      await setContext(data);
    } catch (e) {
      console.error(e);
    }
  };

  const loadedDate =
    cryptoCurrencies && cryptoCurrencies.length > 0 && cryptoCurrencies[0].Date;
  const showInfo = formattedDate(loadedDate) === formattedDate(selectedDate);



  const renderCryptoCurrencies = () => {
    return cryptoCurrencies.map((data, index) => {
      return (
        <tbody key={data._id}>
          <tr>
            <td>{index + 1}</td>
            <td>{data.Currency}</td>
            <td>{"$ " + data.Close}</td>
            <td>{data["24h"]}</td>
            <td>{data["7d"]}</td>
            <td>{data["1m"]}</td>
            <td>{data.Volume}</td>
            <td>{data["Market Cap"]}</td>
          </tr>
        </tbody>
      );
    });
  };

  return (
    <div className="ml-5 mt-5">
      <DatePicker
        className="date-picker"
        selected={selectedDate}
        onChange={(date) => setSelectedDate(date)}
      />
      <button
        onClick={getCryptoCurrencies}
        className="btn btn-info m-5"
        type="button"
      >
        Show Crypto Currencies Data from selected date
      </button>

      {showInfo && (
        <table className="table">
          <thead>
            <tr>
              <th>#</th>
              <th>Coin</th>
              <th>Price</th>
              <th>24 Hours</th>
              <th>7 Days</th>
              <th>1 Month</th>
              <th>24h Volume</th>
              <th>Mkt Cap</th>
            </tr>
          </thead>
          {renderCryptoCurrencies()}
        </table>
      )}
    </div>
  );
};
export default CryptoCurrencies;
