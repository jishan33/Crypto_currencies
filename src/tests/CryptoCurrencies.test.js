import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import CryptoCurrencies from "../components/CryptoCurrencies";
import DatePicker from "react-datepicker";

const flushPromises = () => new Promise(setImmediate);
const waitFor = (milliseconds) =>
  new Promise((resolve) => setTimeout(resolve, milliseconds));

describe("CryptoCurrencies ", () => {
  it("renders CryptoCurrencies component without crashing", () => {
    shallow(<CryptoCurrencies />);
  });

  it("renders DatePicker component without crashing", () => {
    shallow(<DatePicker />);
  });

  test("snapshot renders", () => {
    const component = renderer.create(<CryptoCurrencies />);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
describe("Test Button component", () => {
  it("Test click event", async () => {
    const wrapper = shallow(<CryptoCurrencies />);
    const button = wrapper.find("button");

    expect(wrapper.find("tbody").exists()).toBe(false);

    button.simulate("click");

    await waitFor(200);

    const tables = wrapper.find("tbody");
    expect(tables).toHaveLength(2);

    expect(tables.at(0).text()).toContain("bitcoin");
    expect(tables.at(1).text()).toContain("ethereum");
  });
});
