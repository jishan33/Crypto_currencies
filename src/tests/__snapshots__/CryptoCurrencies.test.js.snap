// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CryptoCurrencies  snapshot renders 1`] = `
<div
  className="ml-5 mt-5"
>
  <div
    className="react-datepicker-wrapper"
  >
    <div
      className="react-datepicker__input-container"
    >
      <input
        className="date-picker"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        readOnly={false}
        type="text"
        value="12/04/2019"
      />
    </div>
  </div>
  <button
    className="btn btn-info m-5"
    onClick={[Function]}
    type="button"
  >
    Show Crypto Currencies Data from selected date
  </button>
</div>
`;
